### DISCUSSION TITLE
Airflow-Azure AD integration always maps to Public role in Airflow

### DISCUSSION THREAD

**Discussion by venkatraghavr** [apache/airflow#31587] [General] [answered]
I am working on Airflow-Azure AD integration. I did all the app registration, app role mapping, user to role mapping, App Permissions etc on AD and I am able to successfully authenticate against AD and get redirected to Airflow page but it then lands me on a page that says -

Your user has no roles and/or permissions!

Unfortunately your user has no roles, and therefore you cannot use Airflow.

Please contact your Airflow administrator ([authentication](https://airflow.apache.org/docs/apache-airflow/stable/security/webserver.html#web-authentication) may be misconfigured) or [log out](https://airflow-d.aks.az.au.singtelgroup.net/logout/) to try again.

I can see successful sign in on AD side and also entry in the Airflow tables ab_users but in ab_user_role table, I see that I am always mapped to 'Public' role. Please find below my Airflow webserver_config.py:

```python
import os
from airflow.configuration import conf
from airflow.utils.log.logging_mixin import LoggingMixin
from flask_appbuilder.security.manager import AUTH_OAUTH
from airflow.www.security import AirflowSecurityManager

SQLALCHEMY_DATABASE_URI = conf.get("database", "SQL_ALCHEMY_CONN")
basedir = os.path.abspath(os.path.dirname(__file__))
CSRF_ENABLED = True
AUTH_TYPE = AUTH_OAUTH

OAUTH_PROVIDERS = [
  { 
        'name':'azure', 
        'token_key':'access_token', 
        'icon':'fa-windows',
        'remote_app': {   
            "api_base_url": "https://login.microsoftonline.com/00000000-0000-0000-0000-000000000",
            "request_token_url": None,
            "request_token_params": {
               "scope": "openid email profile"
            },
            "access_token_url": "https://login.microsoftonline.com/00000000-0000-0000-0000-000000000/oauth2/v2.0/token",
            "access_token_params": {
               "scope": "openid email profile"
            },
            "authorize_url": "https://login.microsoftonline.com/00000000-0000-0000-0000-000000000/oauth2/v2.0/authorize",
            "authorize_params": {
               "scope": "openid email profile"
            },                

            'client_id':'00000000-0000-0000-0000-000000000',
            'client_secret':'00000000-0000-0000-0000-000000000',
            'jwks_uri': 'https://login.microsoftonline.com/common/discovery/v2.0/keys'
        }
  }
]
AUTH_ROLE_ADMIN = 'Admin'
AUTH_USER_REGISTRATION_ROLE = "Public"
AUTH_USER_REGISTRATION = True
AUTH_ROLES_SYNC_AT_LOGIN = True
AUTH_ROLES_MAPPING = {
  "airflow_nonprod_admin": ["Admin"],
  "airflow_nonprod_dev": ["Op"],
  "airflow_nonprod_viewer": ["Viewer"]
}

class AzureCustomSecurity(AirflowSecurityManager, LoggingMixin):
  def get_oauth_user_info(self, provider, response=None):
    if provider == "azure":
        id_token = response["id_token"]
        self.log.debug(str(id_token))
        me = self._azure_jwt_token_parse(id_token)
        return me
    else:
        return {}

SECURITY_MANAGER_CLASS = AzureCustomSecurity```

**Comment by leoliveroso** [comment, marked as answer]
I've same problem

**Reply by Taragolis** [reply]
Unfortunetly "Me to" or "I've same problem" does not much help to identify you problem and help.
At least there is no info about Airflow Version and sample of your web authentifications.
Without this it's unlikely you could find solution of your problem, especially if this problem somehow related to the FAB (Flask-AppBuilder)

**Reply by venkatraghavr** [reply]
I got Airflow-Azure AD Integration code working. Check out the entire working code.

**Comment by venkatraghavr** [comment]
I got it working by replacing above _get_oauth_user_info with below:

def _get_oauth_user_info(self, provider, resp):
            if provider == "azure":
                me = self._azure_jwt_token_parse(resp["id_token"])
                return {
                    "id": me["oid"],
                    "username": me["upn"],
                    "name": me["name"],
                    "email": me["upn"],
                    "first_name": me["given_name"],
                    "last_name": me["family_name"],
                    **"role_keys": me["roles"],**
                }
            return {}
      oauth_user_info = _get_oauth_user_info

What is crucial in above return statement is below, as this "roles" is what maps back to RBAC roles in Airflow:
**"role_keys": me["roles"],**

**Comment by venkatraghavr** [comment]
I got Airflow-Azure AD working using above code. Besides above code in Airflow configuration, required setting inside AAD App registration is absolutely important. You could refer - https://objectpartners.com/2021/12/24/enterprise-auth-for-airflow-azure-ad/

I hope this helps someone.

**Reply by potiuk** [reply]
Since you got it working, maybe @venkatraghavr you could add example of configuring it to Airflow Documentation: similar to https://airflow.apache.org/docs/apache-airflow/stable/security/webserver.html#example-using-team-based-authorization-with-github-oauth -> it should be as easy as clicking the "Suggest a change on this page" button and you should get a PR open where you could add the docs in .rst format. That would be a nice contribution that could help others.

**Reply by venkatraghavr** [reply]
@potiuk I could not get my working example into the Airflow Documentation. Please guide me.

**Reply by potiuk** [reply]
I would have to read the same documentation, I have no idea about Azure-AD, I have no account to test it with, I have no time to individually guess what can be wrong in your configuration that you tried. if you have some concrete problem and you post what you've done and where you are stuck, maybe someone will help you. But bear in mind thi is not free IT support for you. People here help here in their free time. For free. And the best way you can do, is to make them easy to help you by a) explaining what exactly problem you have b) explaining what you 've done already c) explaining what you think happened and what your interpretation is.

Asking "Help me" here does not help anyone to help you. You can also seek for professional paid support if you need someone to guide you.

**Reply by venkatraghavr** [reply]
@potiuk I only requested you for the steps to upload my working code to Airflow documentation (as you suggested) so that it helps someone looking for same solution. I tried out the "Suggest a change on this page" button but got stuck and could not move further. I totally understand that the developer community is supporting a host of free and open-source softwares for the IT industry. Note that I am NOT expecting help or support from anybody here. Thanks.

**Comment by dhruvrkashyap** [comment]
It is due to the variable `AUTH_USER_REGISTRATION_ROLE`
It assign the user who logs in first time the role which you have assigned to this variable.
If changed to `admin`, everyone who log in first time, will be assigned the role of `admin`

You can also manage the role assignment from App Role in Azure AD.
